### PR TITLE
ZK-136: Three status checkbox support

### PR DESCRIPTION
**Pull request by charlesqqq**

### CODE REVIEW

File: zul/src/archive/web/js/zul/wgt/Checkbox.js [hunk 1]
```diff
@@ -229,6 +250,7 @@ zul.wgt.Checkbox = zk.$extends(zul.LabelImageWidget, {
 				checked = real.checked;
 			if (checked != this._checked) { //changed
 				this.setChecked(checked); //so Radio has a chance to override it
+				this.setIndeterminate(false);
```
Comment (wenninghsu) [inline zul/src/archive/web/js/zul/wgt/Checkbox.js]: do we need to set indeterminate to false when checked true -> false ?

File: zktest/test/java/org/zkoss/zktest/zats/test2/F86_ZK_136Test.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		clickAndVerifyLog(isChecked, "false");
+
+		click(jq(".z-checkbox"));
+		waitResponse();
+		clickAndVerifyLog(isIndeterminate, "false");
+		clickAndVerifyLog(isChecked, "true");
+
+		click(toggleIndeterminate);
+		waitResponse();
+		clickAndVerifyLog(isIndeterminate, "true");
```
Comment (scribetw) [inline zktest/test/java/org/zkoss/zktest/zats/test2/F86_ZK_136Test.java]: toggle `Indeterminate` should not affect `checked`

https://www.w3.org/TR/2014/WD-html51-20140617/forms.html#checkbox-state-(type=checkbox)
> The control is never a true tri-state control, even if the element's indeterminate IDL attribute is set to true. The indeterminate IDL attribute only gives the appearance of a third state.

File: zul/src/org/zkoss/zul/Checkbox.java [hunk 3]
```diff
@@ -58,6 +60,32 @@ public Checkbox(String label, String image) {
...
+	 * @return true if checkbox is indeterminate
+	 * @since 8.6.0
+	 */
+	public boolean isIndeterminate() {
+		return _indeterminate;
+	}
+
+	/**
+	 * Set whether checkbox is in indeterminate state,
+	 * if set indeterminate to true, checked will be set to false.
```
Comment (scribetw) [inline zul/src/org/zkoss/zul/Checkbox.java]: This comment is old and should be deleted

File: zul/src/org/zkoss/zul/Checkbox.java [hunk 4]
```diff
@@ -131,6 +159,10 @@ public void setChecked(boolean checked) {
 		if (_checked != checked) {
 			_checked = checked;
 			smartUpdate("checked", _checked);
+			if (!_init && _indeterminate) {
+				_indeterminate = false;
+				smartUpdate("indeterminate", _indeterminate);
```
Comment (scribetw) [inline zul/src/org/zkoss/zul/Checkbox.java]: Just call `setIndeterminate(false)`?

File: zul/src/org/zkoss/zul/Checkbox.java [hunk 5]
```diff
@@ -131,6 +159,10 @@ public void setChecked(boolean checked) {
 		if (_checked != checked) {
 			_checked = checked;
 			smartUpdate("checked", _checked);
+			if (!_init && _indeterminate) {
```
Comment (scribetw) [inline zul/src/org/zkoss/zul/Checkbox.java]: Add javadoc that changing checked will set indeterminate false

File: zul/src/org/zkoss/zul/Checkbox.java [hunk 6]
```diff
@@ -58,6 +60,32 @@ public Checkbox(String label, String image) {
 		super(label, image);
 	}
 
+	/**
+	 * Return whether checkbox is in indeterminate state,
+	 * if indeterminate is true, checked must be false.
```
Comment (scribetw) [inline zul/src/org/zkoss/zul/Checkbox.java]: This comment is old and should be deleted

File: zul/src/archive/web/js/zul/wgt/Checkbox.js [hunk 7]
```diff
@@ -173,7 +173,28 @@ zul.wgt.Checkbox = zk.$extends(zul.LabelImageWidget, {
...
+		 * Return whether checkbox is in indeterminate state,
+		 * if indeterminate is true, checked must be false.
+		 * Default: false.
+		 *
+		 * @return boolean
+		 * @since 8.6.0
+		 */
+		/**
+		 * Set whether checkbox is in indeterminate state,
+		 * if set indeterminate to true, checked state will be set to false.
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/wgt/Checkbox.js]: This comment is old and should be deleted

**Comment by wenninghsu** [comment]
@zkadmin rebase

**Comment by zkadmin** [comment]
I just pushed up the changes, enjoy!

**Pull request merged**